Warn when a `#[test]`-like built-in attribute macro is present multiple times.

Ethiraric opened the pull request:
Fixes #90979.

GuillaumeGomez commented:
r? @petrochenkov

Milo123459 commented:
Shouldn't this emit an error, not a warning? I'm trying to think of a use case where having multiple `#[test]`'s is a good idea and would probably make sense for it to be an error. Just giving my opinion though, it might make sense to have it as a warning.

Ethiraric commented:
It could very well be an error. I don't see any case either where this would be the expected behavior. Having a hard error on that is maybe too inconvenient on what was probably an oversight from the developer and could break existing codebases.

Milo123459 commented:
> It could very well be an error. I don't see any case either where this would be the expected behavior. Having a hard error on that is maybe too inconvenient on what was probably an oversight from the developer and could break existing codebases.

Implementing a fix for it would be super easy; it just has to remove the other usages of #[test] on that function.

Ethiraric commented:
How does that weigh in whether this should be an error or a warning?

GuillaumeGomez commented:
It's up to @rust-lang/compiler team I guess. But in any case, I think we have to first start with a warning saying it'll become a hard error in the future if we decide that it's not supposed to compile.

Milo123459 commented:
> How does that weigh in whether this should be an error or a warning?

Read it wrong. My bad.

I agree - throwing a hard error might be inconsistent.

lcnr commented:
we currently allow the following without any warnings as well
```rust
#[inline]
#[inline]
fn foo() {}
```
I think that ideally we should hard error for duplicate builtin attributes though starting with a future incompat lint is probably the right choice :thinking: imo we should start a crater run for this, you would have to first convert the warning to an error in that case.

nagisa commented:
Right now the story around duplicate attributes is definitely somewhat erratic. There are some attributes which when duplicated will implement their function twice and sometimes cause an error (though outside of the attribute system). `derive` and `test` are an example of this. Some others will behave as if the attribute was specified once (e.g. `#[no_mangle]`), and some, such as `inline` will have an overriding behaviour (where only the last applies).

To be fair, I'm not sure if `test` having the “repeated operation” is the right model for `#[test]` in the first place – I would've expected it to work more like `no_mangle`, in which case warning would not be necessary either.

There are also considerations with regards to e.g. macros, where ability to duplicate some attributes may make things significantly easier for them.

GuillaumeGomez commented:
> There are also considerations with regards to e.g. macros, where ability to duplicate some attributes may make things significantly easier for them.

This is mostly why I think a warning should be enough.

Ethiraric commented:
> you would have to first convert the warning to an error in that case.

As implemented in this PR, the diagnostic is reported as a warning.

Should you decide on what the behavior for repeated builtin attributes should be, I am willing, in other PRs, to change their behaviors according to what you think is best.

In any case, let me know what changes you want me to do in this PR.

ehuss commented:
> we currently allow the following without any warnings as well

Duplicate attributes should now warn on the latest nightly.

Ethiraric commented:
> Duplicate attributes should now warn on the latest nightly.

It doesn't as of :
```
rustc 1.58.0-nightly (65c55bf93 2021-11-23)
binary: rustc
commit-hash: 65c55bf931a55e6b1e5ed14ad8623814a7386424
commit-date: 2021-11-23
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

ehuss commented:
@Ethiraric Can you say more about what you tried?  It should display a warning like this:

```
warning: unused attribute
 --> src/lib.rs:2:1
  |
2 | #[inline]
  | ^^^^^^^^^ help: remove this attribute
  |
  = note: `#[warn(unused_attributes)]` on by default
note: attribute also specified here
 --> src/lib.rs:1:1
  |
1 | #[inline]
  | ^^^^^^^^^
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
```

If you tried with `#[test]`, then that is because the lint intentionally doesn't fire on proc-macros.

Ethiraric commented:
I tried to compile the file I linked on the issue with the compiler I posted the version of.

Can you try to compile the snippet from #90979 and tell me whether you have a warning?

ehuss commented:
> Why is that that `#[test]` doesn't show any diagnostic?

`#[test]` is treated like a proc-macro, and not a built-in attribute.  The lint only fires for built-in attributes.

Ethiraric commented:
But, with the current behaviour, I don't see any case annotating an item more than once with `#[test]` would be desired. This is why I thought a warn (not a hard error) would be nice. 

I get that because it's a particular macro there currently is no warn. I have implemented this diagnostic here because I think there should be at least something that warns user. 

Do you think this would be a bad addition?

GuillaumeGomez commented:
I think the same as @Ethiraric in here: having a test run as many times as the `#[test]` attribute is present deserves at least a warning so at least you know why it's happening (even more if it comes from a macro!).

ehuss commented:
Sorry, I didn't mean to mislead. I wasn't advocating against this change. I was just pointing out in the discussion above that things had changed (for built-in attribute) as of 2 days ago, and people didn't seem to be aware.

GuillaumeGomez commented:
Funny that @Ethiraric came across this just in the middle of this "change wave". :)

petrochenkov commented on compiler/rustc_builtin_macros/src/test.rs:
```diff
@@ -80,7 +114,17 @@ pub fn expand_test_or_bench(
+        Annotatable::Item(i) => {
+            if let Some(attr) = i.attrs.iter().filter(|attr| is_test_attribute(attr)).next() {
+                let sd = &cx.sess.parse_sess.span_diagnostic;
+                sd.struct_span_warn(
```
Hard-coded warnings should almost never be used, either lints or errors are supposed to be used instead.
I need to think what is a better course of action in this specific case.

petrochenkov replied:
We cannot recognize other calls to `#[test]` precisely here due to cases like this
```rust
#[test] // built-in `test`
#[arbitrary_attribute]
#[test] // in theory this can resolve to a different non-builtin macro attribute
        // or this can be a significant input to `#[arbitrary_attribute]`
```
, so if we error on this it should be a documented part of the `test`'s behavior.

petrochenkov replied:
TBH, doing nothing and closing #90979 as not-an-issue would also be a good solution.

petrochenkov replied:
I suggest adding a new warn-by-default lint called `duplicate_macro_attributes`, and reuse the wording from https://github.com/rust-lang/rust/pull/88681.

petrochenkov commented:
The full list of `test`-like built-in attribute macros is `cfg_eval`, `global_allocator`, `test_case`, `test` and `bench`.
(They can be found by searching for calls to `check_builtin_macro_attribute`.)

Ethiraric commented:
Thank you for your review and pointers. Should I add the lint to all built-in attribute macros in this PR or open a new one?

petrochenkov commented:
>Should I add the lint to all built-in attribute macros in this PR or open a new one?

I thinks it's better to do it in this PR.

Ethiraric commented:
Hi! I'm apologize for taking so long to get back to you on that PR.

I have added the lint on `test`, `bench`, `cfg_eval` and `test_case`.
I tried adding it to `global_allocator`, but duplicating the attribute causes an error already (`cannot define multiple global allocators`).

The tests do not include `test_case`. I haven't found a way of making the code compile, taking as an example [custom_test_frameworks/full.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/custom_test_frameworks/full.rs).

I am unsure about whether my approach to firing the lint is good. Please tell me if the function I added in `session.rs` needs any change. The code is largely inspired from [`rustc_middle::lint::struct_lint_level`](https://github.com/rust-lang/rust/blob/9c83f8c4d1e562cd73a29a00828d68e2b520a086/compiler/rustc_middle/src/lint.rs#L205).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4527616095?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/762b1a56aff12b1a889e5f795cd80a29983ab017/checks/4527616095/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling serde v1.0.125
   Compiling serde_json v1.0.59
   Compiling lint-docs v0.1.0 (/checkout/src/tools/lint-docs)
    Finished release [optimized] target(s) in 6.37s
warning: the code example in lint `duplicate_macro_attributes` in /checkout/compiler/rustc_lint_defs/src/builtin.rs failed to generate the expected output: did not find lint `duplicate_macro_attributes` in output of example, got:

Rustbook (x86_64-unknown-linux-gnu) - rustc
Set({"src/tools/cargo/src/doc"}) not skipped for "bootstrap::doc::CargoBook" -- not in ["src/tools/tidy"]
Rustbook (x86_64-unknown-linux-gnu) - cargo
---
doc tests for: /checkout/src/doc/rustc/src/what-is-rustc.md
Set({"src/tools/lint-docs"}) not skipped for "bootstrap::test::LintDocs" -- not in ["src/tools/tidy"]
 finished in 1.071 seconds
Generating lint docs (x86_64-unknown-linux-gnu)
error: failed to test example in lint docs for `duplicate_macro_attributes` in /checkout/compiler/rustc_lint_defs/src/builtin.rs:3633: did not find lint `duplicate_macro_attributes` in output of example, got:




command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/lint-docs" "--src" "/checkout/compiler" "--out" "/checkout/obj/build/x86_64-unknown-linux-gnu/md-doc/rustc/src/lints" "--rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustc-target" "x86_64-unknown-linux-gnu" "--validate"
This error was generated by the lint-docs tool.
This error was generated by the lint-docs tool.
This tool extracts documentation for lints from the source code and places


them in the rustc book. See the declare_lint! documentation
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/macro.declare_lint.html


To re-run these tests, run: ./x.py test --keep-stage=0 src/tools/lint-docs
The --keep-stage flag should be used if you have already built the compiler
Build completed unsuccessfully in 0:36:38
and are only modifying the doc comments to avoid rebuilding the compiler.

```

</details>

Ethiraric commented:
Updated the comment for the lint.

I think however the build fails because the snippet of code should be run with `--test`. I know not how to instruct the compiler to do so.

petrochenkov commented on compiler/rustc_builtin_macros/src/util.rs:
```diff
@@ -10,3 +12,28 @@ pub fn check_builtin_macro_attribute(ecx: &ExtCtxt<'_>, meta_item: &MetaItem, na
+
+/// Emit a warning if the item is annotated with the given attribute. This is used to diagnose when
+/// an attribute may have been mistakenly duplicated.
+pub fn warn_on_duplicated_attribute(session: &Session, item: &Annotatable, name: Symbol) {
```
```suggestion
pub fn warn_on_duplicate_attribute(session: &Session, item: &Annotatable, name: Symbol) {
```
https://english.stackexchange.com/questions/67463/duplicate-data-or-duplicated-data

petrochenkov commented on compiler/rustc_session/src/session.rs:
```diff
@@ -321,6 +325,23 @@ impl Session {
     ) -> DiagnosticBuilder<'_> {
         self.diagnostic().struct_span_warn_with_code(sp, msg, code)
     }
+    pub fn span_lint<S: Into<MultiSpan>>(&self, sp: S, lint: &'static Lint) {
```
This shouldn't be necessary, see how `buffer_lint` is used in other built-in macros for reporting e.g. `BAD_ASM_STYLE`.

(Also, this is not a future-compatibility lint.)

petrochenkov commented on compiler/rustc_lint_defs/src/builtin.rs:
```diff
@@ -3629,3 +3629,32 @@ declare_lint! {
+    pub DUPLICATE_MACRO_ATTRIBUTES,
+    Warn,
+    "duplicated attribute"
+}
```
This lint needs to be added to the `HardwiredLints` list in the same file.

petrochenkov commented:
>I think however the build fails because the snippet of code should be run with `--test`. I know not how to instruct the compiler to do so.
You can use
```
```rust,ignore
```
for disabling the `lint-docs` check.

Ethiraric commented:
Thanks for your review. I should have fixed all your requested changes!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4540620272?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6fb3ca256dec484310c23602325e3cc1bc66d440/checks/4540620272/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 628 error codes
* highest error code: E0786
tidy error: /checkout/compiler/rustc_lint_defs/src/builtin.rs:3615: unexplained "```ignore" doctest; try one:

* make the test actually pass, by adding necessary imports and declarations, or
* use "```text", if the code is not Rust code, or
* use "```compile_fail,Ennnn", if the code is expected to fail at compile time, or
* use "```should_panic", if the code is expected to fail at run time, or
* use "```no_run", if the code should type-check but not necessary linkable/runnable, or
* explain it like "```ignore (cannot-test-this-because-xxxx)", if the annotation cannot be avoided.

Found 502 error codes
Found 0 error codes with no tests
Done!
Done!
* 361 features
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:11
```

</details>

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 2be94d43019961f855589c1c6e2e27063537d30a has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 2be94d43019961f855589c1c6e2e27063537d30a -->
<!-- homu: {"type":"Approved","sha":"2be94d43019961f855589c1c6e2e27063537d30a","approver":"petrochenkov"} -->
